README.rst
=======
mychevy
=======


.. image:: https://img.shields.io/pypi/v/mychevy.svg
        :target: https://pypi.python.org/pypi/mychevy

.. image:: https://img.shields.io/travis/sdague/mychevy.svg
        :target: https://travis-ci.org/sdague/mychevy

.. image:: https://readthedocs.org/projects/mychevy/badge/?version=latest
        :target: https://mychevy.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/sdague/mychevy/shield.svg
     :target: https://pyup.io/repos/github/sdague/mychevy/
     :alt: Updates


Python interface to My Chevy website

Unlike Tesla, GM does not provide a consumer level API to their vehicles. I
tried to sign up for their developer program after purchasing my Chevy Bolt,
but so far it's all been black holed. They do provide a useful My Chevy
website, where you can log in with your OnStar credentials and see things like
how charged your battery is. This is all built with a javascript framework, and
the data loads off the OnStar network with a 60 - 120 second delay (OnStar is a
rather slow proprietary cellular network)

This library logs into the mychevy website, and uses the API that the javascript
on the site uses to load the onstar data.

Installation
============

Installation for this library is just a pip install.

1. pip install mychevy


Usage
=====

Usage is very basic.

.. code-block:: python

   from mychevy.mychevy import MyChevy

   page = MyChevy("<username>", "<password>")
   # This takes up to 2 minutes to return, be patient

   # build credentials and starts a session on the mychevy site
   page.login()

   # gets a list of cars associated with your account
   page.get_cars()

   # update stats for all cars, and returns the list of cars. This will take
   # 60+ seconds per car
   cars = page.update_cars()

   # Percent battery charge
   print(cars[0].percent)


Every invocation of ``login()`` creates a whole new session to avoid
credential timeouts.

It is not recommended that you run this very frequently. Something like once an
hour will give you basic data, and shouldn't overload anyone's systems.

Testing
=======

Because there are so many ways this can go wrong, a basic cli tool has been
provided.

.. code-block:: bash

   > mychevy -c config.ini
   Loading data, this takes up to 2 minutes...
   <EVCar vin="...", range=185 miles, bat=100%, plugged_in=True, mileage=903 miles, charging=Your battery is fully charged., charge_mode=Departure Based, eta=None, state="">

config.ini must include your user and password for the mychevy site in the
following format:

.. code-block:: ini

   [default]
   user = my@email.address
   passwd = my@wes0mepa55w0rd


Caveats
=======

There are so many caveats.... This software aspires to be the gloriously robust
bubble gum and duct tape of which it has heard makes the internet go round.

* JSON formats are guessed at

  The use of the sessions capture and transfer, and inspecting json returned
  still creates slightly different parameters than are used by the website. The
  set of keys and values are guessed at. It's all kind of fragile and
  heuristic.

* The MyChevy website OnStar link is not robust

  In the first month with the Bolt I've seen two multi hour outages of the
  mychevy website being able to connect to their OnStar backend gateway. One
  lasted a whole day. The OnStar link from the Android App worked fine during
  these windows of time. So it's not an OnStar failure, but it's a lack of
  robustness somewhere on the Web side, or the gateway dedicated for serving
  OnStar requests.

As such, this software will always be classified Alpha on Pypi. It can and will
break. For that I'm sorry. But it's the best I've got.


* Free software: Apache Software License 2.0
* Documentation: https://mychevy.readthedocs.io.


Features
--------

* TODO

Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

2.1.1 (2021-01-04)
------------------
* Fix tests

2.1.0 (2021-01-04)
------------------
* Add non EV attributes
* Fix UA string

2.0.1 (2020-04-21)
------------------
* flake8 fixes

2.0.0 (2020-04-21)
------------------
* Support new Azure based login. Thanks to @jaygun and me@lukecyca.com for core logic. Fixes issue #11

1.2.0 (2019-01-02)
------------------
* Support CA urls
* Reduce implicit retries

1.1.0 (2018-12-13)
------------------
* add explicit charging property

1.0.0 (2018-12-03)
------------------
* remove the selenium requirement (thanks starbuddy1995@gmail.com)
* get my.chevy working with the new website (thanks starbuddy1995@gmail.com)

0.4.0 (2018-05-09)
------------------
* my.chevy website is finally back up, data structure parsing adjusted to match it

0.3.2 (2018-02-04)
------------------

* Add simple unit testing
* Stop testing on python 3.4 so we can use more advanced exceptions

0.3.1 (2018-01-31)
------------------

* Nicer error handling when server throws an error

0.3.0 (2018-01-30)
------------------

* Change data collection to use json parsing instead of screen scraping after login

0.2.0 (2018-01-21)
------------------

* Upgraded cli to use `click`

0.1.0 (2017-12-31)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for mychevy."""


tests/test_mychevy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `mychevy` package."""

import unittest

import pytest

from mychevy.mychevy import EVCar, ServerError

CAR1 = {
    "vin": "fakevin",
    "vehicle_id": "123",
    "onstarAccountNumber": "123",
    "year": "2017",
    "make": "Chevy",
    "model": "Bolt",
    "imageUrl": ""
}

# DATA packet:
PKT1 = b'{"messages":[],"serverErrorMsgs":[],"data":{"dataAsOfDate":1516671611000,"batteryLevel":70,"chargeState":"not_charging","plugState":"plugged","rateType":"PEAK","voltage":240,"electricRange":132,"totalRange":132,"chargeMode":"DEPARTURE_BASED","electricMiles":1601,"gasMiles":0,"totalMiles":1601,"percentageOnElectric":1,"fuelEconomy":1000,"electricEconomy":45,"combinedEconomy":11,"fuelUsed":132,"electricityUsed":132,"estimatedGallonsFuelSaved":61.13,"estimatedCO2Avoided":1185.92,"estimatedFullChargeBy":"5:00 a.m."}}'  # noqa


class TestMyChevy(unittest.TestCase):

    def test_car_parse(self):
        car = EVCar(CAR1)
        assert car.vid == "123"

    def test_car_server_error(self):
        car = EVCar(CAR1)
        with pytest.raises(ServerError):
            car.from_json(
                b'{"messages": [], "serverErrorMsgs": [], '
                b'"data": "SERVER ERROR"}')

    def test_json_parse(self):
        car = EVCar(CAR1)
        car.from_json(PKT1)

        assert car.plugged_in is True
        assert car.batteryLevel == 70
        assert car.voltage == 240
        assert car.totalRange == 132
        assert car.totalMiles == 1601
        assert car.estimatedFullChargeBy == "5:00 a.m."


mychevy/cli.py
# -*- coding: utf-8 -*-

"""Console script for mychevy."""

import configparser

import click

from mychevy.mychevy import MyChevy

CONTEXT_SETTINGS = dict(help_option_names=['-h', '--help'])


@click.command(context_settings=CONTEXT_SETTINGS)
@click.option('--config', '-c', type=click.File('r'),
              required=True,
              help="Config file with my.chevy credentials")
@click.option('--show-browser', '-S', is_flag=True,
              help="Show browser window when running")
def main(config=None, show_browser=None):
    """Console script for mychevy"""
    cfile = configparser.ConfigParser()
    cfile.read_file(config)

    page = MyChevy(cfile["default"]["user"], cfile["default"]["passwd"])
    click.echo("Loading data, this takes up to 2 minutes...")
    page.login()
    page.get_cars()
    cars = page.update_cars()
    for c in cars:
        click.echo(c)


if __name__ == "__main__":
    main()


mychevy/debug.py
# -*- coding: utf-8 -*-

"""Console script for mychevy."""

import configparser
import logging

import click

from mychevy.mychevy import MyChevy, ServerError

CONTEXT_SETTINGS = dict(help_option_names=['-h', '--help'])


@click.command(context_settings=CONTEXT_SETTINGS)
@click.option('--config', '-c', type=click.File('r'),
              required=True,
              help="Config file with my.chevy credentials")
@click.option('--verbose', '-v', default=False, is_flag=True,
              help="Run more verbose")
def main(config=None, verbose=False):
    """Console script for mychevy"""
    cfile = configparser.ConfigParser()
    cfile.read_file(config)

    if verbose:
        logging.basicConfig(level=logging.DEBUG)

    page = MyChevy(cfile["default"]["user"], cfile["default"]["passwd"])
    click.echo("Logging in... this takes a bit")
    page.login()
    page.get_cars()
    click.echo("Displaying found cars")
    for c in page.cars:
        click.echo(c)
    click.echo("Updating cars with data")
    try:
        if cfile.has_option("default","vin"):
            page.update_cars(cfile["default"]["vin"])
        else:
            page.update_cars()

        click.echo("Displaying found cars with data")
        for c in page.cars:
            click.echo(c)
    except ServerError as e:
        click.echo("OnStar Network Failure: %s" % e)


if __name__ == "__main__":
    main()


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file."""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except ImportError:
    from urllib.request import urlopen


GITHUB_REPO = 'sdague/mychevy'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key.

    Work around keys with incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning."""
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    """Load yaml config file at the given path."""
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    """Save yaml config file at the given path."""
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Put `encrypted_password` into the deploy section of .travis.yml."""
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    """Add a PyPI password to .travis.yml so that Travis can deploy to PyPI.

    Fetch the Travis public key for the repo, and encrypt the PyPI password
    with it before adding, so that only Travis can decrypt and use the PyPI
    password.
    """
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


mychevy/mychevy.py
# Copyright 2017 Sean Dague
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

"""Main module."""

from functools import wraps
import json
import logging
import re
import time
import urllib

import requests

_LOGGER = logging.getLogger(__name__)

TIMEOUT = 120
KEY = 15258643512041

LOGIN_URL = "https://custlogin.gm.com/gmb2cprod.onmicrosoft.com/B2C_1A_SeamlessMigration_SignUpOrSignIn/SelfAsserted?tx={}&p=B2C_1A_SeamlessMigration_SignUpOrSignIn"  # noqa
TOKEN_URL = "https://custlogin.gm.com/gmb2cprod.onmicrosoft.com/B2C_1A_SeamlessMigration_SignUpOrSignIn/api/CombinedSigninAndSignup/confirmed?csrf_token={}&tx={}&p=B2C_1A_SeamlessMigration_SignUpOrSignIn"  # noqa

URLS = {
    "us": {
        "success": "https://my.chevrolet.com/init/loginSuccessData",
        "oc_login": "https://my.chevrolet.com/oc_login",
        "loginSuccessData": "https://my.chevrolet.com/api/init/loginSuccessData",
        "home": "https://my.chevrolet.com/home",
        "login": "https://my.chevrolet.com/oc_login",
        "evstats": (
            "https://my.chevrolet.com/api/vehicleProfile/"
            "{0}/{1}/evstats/false?cb={2}.{3}"
        ),
        "session": (
            "https://my.chevrolet.com/vehicleProfile/"
            "{0}/{1}/createAppSessionKey?cb={2}.{3}"
        ),
    },
    "ca": {
        "success": "https://my.gm.ca/chevrolet/en/init/loginSuccessData",
        "home": "https://my.gm.ca/gm/en/home",
        "oc_login": "https://my.gm.ca/gm/en/oc_login",
        "loginSuccessData": "https://my.gm.ca/gm/en/api/init/loginSuccessData",
        "login": "https://my.gm.ca/chevrolet/en/oc_login",
        "evstats": (
            "https://my.gm.ca/chevrolet/en/api/vehicleProfile/"
            "{0}/{1}/evstats/false?cb={2}.{3}"
        ),
        "session": (
            "https://my.gm.ca/chevrolet/en/api/vehicleProfile/"
            "{0}/{1}/createAppSessionKey?cb={2}.{3}"
        ),
    },
}


def get_url(kind="", country="us"):
    return URLS[country][kind]


settings_json_re = re.compile("var SETTINGS = ({.*})")
id_token_re = re.compile("name='id_token'.*value='(.*)'/>")


USER_AGENT = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) "
    "AppleWebKit/537.36 (KHTML, like Gecko) "
    "Chrome/75.0.3770.80 Safari/537.36"
)

# b'{"messages":[],"serverErrorMsgs":[],"data":{"dataAsOfDate":1517317946000,"batteryLevel":82,"chargeState":"not_charging","plugState":"unplugged","rateType":"PEAK","voltage":0,"electricRange":155,"totalRange":182,"chargeMode":"DEPARTURE_BASED","electricMiles":1843,"gasMiles":0,"totalMiles":1843,"percentageOnElectric":1,"fuelEconomy":1000,"electricEconomy":44,"combinedEconomy":9,"fuelUsed":182,"electricityUsed":182,"estimatedGallonsFuelSaved":70.45,"estimatedCO2Avoided":1366.73,"estimatedFullChargeBy":"5:00
# a.m."}}'


class ServerError(Exception):
    pass


CAR_ATTRS = (
    "chargeMode",
    "chargeState",
    "batteryLevel",
    "electricRange",
    "totalRange",
    "totalMiles",
    "electricMiles",
    "gasRange",
    "gasFuelLevelPercentage",
    "fuelEconomy",
    "gasMiles",
    "voltage",
    "estimatedFullChargeBy",
)


def retry(exceptions, tries=3, delay=3, backoff=2, logger=None):
    """
    Retry calling the decorated function using an exponential backoff.

    Args:
        exceptions: The exception to check. may be a tuple of
            exceptions to check.
        tries: Number of times to try (not retry) before giving up.
        delay: Initial delay between retries in seconds.
        backoff: Backoff multiplier (e.g. value of 2 will double the delay
            each retry).
        logger: Logger to use. If None, print.
    """

    def deco_retry(f):
        @wraps(f)
        def f_retry(*args, **kwargs):
            mtries, mdelay = tries, delay
            while mtries > 1:
                try:
                    return f(*args, **kwargs)
                except exceptions as e:
                    msg = "{}, Retrying in {} seconds...".format(e, mdelay)
                    if logger:
                        logger.warning(msg)
                    else:
                        print(msg)
                    time.sleep(mdelay)
                    mtries -= 1
                    mdelay *= backoff
            return f(*args, **kwargs)

        return f_retry  # true decorator

    return deco_retry


class EVCar(object):
    def __init__(self, car):
        super(EVCar, self).__init__()
        self.vin = car["vin"]
        self.vid = car["vehicle_id"]
        self.onstar = car["onstarAccountNumber"]
        self.year = car["year"]
        self.make = car["make"]
        self.model = car["model"]
        self.img = car["imageUrl"]

        # computed binaries
        self.plugged_in = False

        # car stats that we'll update later
        self.chargeMode = ""
        self.fuelEconomy = 0
        self.batteryLevel = ""
        self.chargeState = ""
        self.electricRange = 0
        self.totalRange = 0
        self.totalMiles = 0
        self.electricMiles = 0
        self.gasRange = 0
        self.gasFuelLevelPercentage = 0
        self.gasMiles = 0
        self.voltage = 0
        self.estimatedFullChargeBy = ""

    @property
    def name(self):
        return "{0} {1} {2}".format(self.year, self.make, self.model)

    @property
    def charging(self):
        return self.chargeState == "charging"

    def update(self, *args, **kwargs):
        for k, v in kwargs.items():
            if hasattr(self, k):
                setattr(self, k, v)
            else:
                raise KeyError("No attribute named %s" % k)

    def from_json(self, data):
        try:
            res = json.loads(data.decode("utf-8"))

            # I've never actually seen serverErrorMsgs, but allow for them
            if res["serverErrorMsgs"] or type(res["data"]) == str:
                raise ServerError(res)

            d = res["data"]

            self.plugged_in = d["plugState"] == "plugged"
            _LOGGER.debug("Data: {}".format(d))
            for a in CAR_ATTRS:
                # if the attr exists, set it
                if a in d:
                    setattr(self, a, d[a])

        except json.JSONDecodeError:
            _LOGGER.exception("Failure to decode json: %s" % data)
        except KeyError:
            _LOGGER.exception("Expected key not found")

    def __str__(self):
        return (
            "<EVCar name=%s, electricRange=%s miles, batteryLevel=%s%%, "
            "gasRange=%s miles, fuelEconomy=%s mpg, gasFuelLevelPercentage=%s%%, "
            "plugged_in=%s, "
            "totalMiles=%s miles, chargeState=%s, chargeMode=%s, "
            "estimatedFullChargeBy=%s>"
            % (
                self.name,
                self.electricRange,
                self.batteryLevel,
                self.gasRange,
                self.fuelEconomy,
                self.gasFuelLevelPercentage,
                self.plugged_in,
                self.totalMiles,
                self.chargeState,
                self.chargeMode,
                self.estimatedFullChargeBy,
            )
        )


class MyChevy(object):
    def __init__(self, user, passwd, country="us"):
        super(MyChevy, self).__init__()

        self.user = user
        self.passwd = passwd
        self.cars = []
        self.cookies = None
        self.history = []
        self.session = None
        self.account = None
        self.country = country

    def login(self):
        """New login path, to be used with json data path."""
        # Get the main page
        self.session = requests.Session()

        # It doesn't like an empty session so load the login page first.
        r = self.session.get(get_url("home", self.country), timeout=TIMEOUT)
        initial_url = urllib.parse.urlparse(r.request.url)
        nonce = urllib.parse.parse_qs(initial_url.query).get("nonce")[0]

        _LOGGER.debug("Initial URL %s, Nonce %s", initial_url, nonce)
        m = settings_json_re.search(r.text)
        if not m:
            raise ValueError("SETTINGS not found in response")

        settings_json = json.loads(m[1])
        csrf = settings_json["csrf"]
        trans_id = settings_json["transId"]

        _LOGGER.debug(
            "Settings %s, CSRF %s, Trans_id %s", settings_json, csrf, trans_id
        )

        # Login Request
        r = self.session.post(
            LOGIN_URL.format(trans_id),
            {
                "request_type": "RESPONSE",
                "logonIdentifier": self.user,
                "password": self.passwd,
            },
            headers={
                "X-CSRF-TOKEN": csrf,
            },
        )

        # Generate Auth Code and ID Token
        r = self.session.get(TOKEN_URL.format(csrf, trans_id))
        r.raise_for_status()
        _LOGGER.debug("ID Token Content: %s", r.content)
        m = id_token_re.search(r.text)
        if not m:
            raise ValueError("id_token not found in response")

        id_token = m.group(1)

        # Post ID Token
        r = self.session.post(
            get_url("oc_login", self.country),
            {"id_token": id_token},
        )
        r.raise_for_status()
        self.account = self.session.get(
            get_url("loginSuccessData", self.country), timeout=TIMEOUT
        )

    def get_cars(self):
        try:
            data = json.loads(self.account.content.decode("utf-8"))
            if data["serverErrorMsgs"]:
                raise Exception(data["serverErrorMsgs"])

            self.cars = []
            _LOGGER.debug("Vehicles: %s", data["data"]["vehicleMap"])
            for vid, vehicle in data["data"]["vehicleMap"].items():
                self.cars.append(EVCar(vehicle))
        except Exception:
            raise Exception(
                """
Something went wrong!

Cookies: %s

Content: %s

Location: %s
            """
                % (self.account.cookies, self.account.content, self.account.history)
            )

    @retry(ServerError, logger=_LOGGER)
    def _fetch_car(self, car):
        headers = {"user-agent": USER_AGENT}
        _LOGGER.debug("Fetching car...")
        now = int(round(time.time() * 1000))
        session = get_url("session", self.country).format(car.vin, car.onstar, now, KEY)
        res = self.session.get(
            session,
            headers=headers,
            cookies=self.cookies,
            allow_redirects=False,
            timeout=TIMEOUT,
        )

        now = int(round(time.time() * 1000))
        url = get_url("evstats", self.country).format(car.vin, car.onstar, now, KEY)
        res = self.session.get(
            url,
            headers=headers,
            cookies=self.cookies,
            allow_redirects=False,
            timeout=TIMEOUT,
        )

        _LOGGER.debug("Vehicle data: %s" % res.content)
        car.from_json(res.content)

    def update_cars(self):
        for c in self.cars:
            self._fetch_car(c)
        return self.cars
